pkg/clusteragent/autoscaling/controller.go

```diff
@@ -0,0 +1,148 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build kubeapiserver
+
+package autoscaling
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+)
+
+// Processor defines the interface that needs to be implemented by the controller
+type Processor interface {
+	Process(key, ns, name string)
+}
+
+// Controller is generic controller that can be used to implement a controller for a specific resource
+type Controller struct {
+	processor Processor
+	synced    cache.InformerSynced
+	context   context.Context
+
+	// Fields available to child controllers
+	ID        string
+	Client    dynamic.Interface
+	Lister    cache.GenericLister
+	Workqueue workqueue.RateLimitingInterface
```

Comment:
Do we need these fields to be public ?

Reply:
This is a generic controller, so yes we need to expose them so that they can be accessed by controllers embedding it.
We could also move them to a `ControllerContext` and only allow to get that.